File: src/Anf.Engine.Easy/Downloading/DownloadBox.cs
using System;
using System.Threading;

namespace Anf.Easy.Downloading
{
    public class DownloadBox
    {
        public DownloadBox(DownloadTask task, DownloadLink link)
        {
            Task = task;
            Link = link;
        }
        internal CancellationTokenSource TokenSource { get; set; }

        public event Action<DownloadBox> Canceled;

        public string Address => Link.Request?.Entity?.ComicUrl;

        public DownloadTask Task { get; }

        public DownloadLink Link { get; }

        public void Cancel()
        {
            var tsk = TokenSource;
            if (tsk != null && !tsk.IsCancellationRequested)
            {
                tsk.Cancel();
                Canceled?.Invoke(this);
            }
        }
    }
}

File: anf-rs/crates/anf-easy/src/downloading/download_box.rs
use crate::downloading::download_task::DownloadTask;
use tokio_util::sync::CancellationToken;

pub struct DownloadBox {
    pub task: DownloadTask,
    pub address: String,
    token_source: CancellationToken,
}

impl DownloadBox {
    pub fn new(task: DownloadTask, address: String) -> Self {
        let token_source = task.token.clone();
        Self {
            task,
            address,
            token_source,
        }
    }
    pub fn cancel(&self) {
        self.token_source.cancel();
    }
    pub fn is_cancelled(&self) -> bool {
        self.token_source.is_cancelled()
    }
}
